contracts/staking/fuse/FuseStaking.sol

```diff
@@ -0,0 +1,743 @@
+	}
+
+	function _getLastTotalFinalizedSupply() internal view returns(uint256) {
+		return globalStakingVariablesHistory.length > 0 ? globalStakingVariablesHistory[globalStakingVariablesHistory.length - 1].totalFinalizedSupply : 0;
+	}
+
+	function _getRewardPerTokenPerUser(address _account) internal view returns(uint256) {
+		if (collectUBIInterestCallTimes.length > 1) {
+			uint256 rewardPerTokenPerUser = 0;
+			for (uint256 i = 1; i < collectUBIInterestCallTimes.length; i++) {
+				if (collectUBIInterestCallTimeIdxToUserStakeTime[i][_account] == 0) continue;
+				rewardPerTokenPerUser +=
+					_rewardPerToken(
+						globalStakingVariablesHistory[i].rewardPerToken,
+						collectUBIInterestCallTimes[i] - collectUBIInterestCallTimeIdxToUserStakeTime[i - 1][_account],
+						globalStakingVariablesHistory[i].rewardRate,
+						globalStakingVariablesHistory[i].totalFinalizedSupply
+					);
+			}
+			return rewardPerTokenPerUser;
+		} else {
+			return 0;
+		}
+	}
+
+	function earned(address account) public view returns (uint256) {
+		return pendingStakes[account] * (_getRewardPerTokenPerUser(account) - userRewardPerTokenPaid[account]) / PRECISION;
+	}
+
+	function _getReward(address _to) internal {
+		uint256 reward = earned(_to);
+		if (reward > 0) {
+				userRewardPerTokenPaid[_to] = _getRewardPerTokenPerUser(_to);
+				payable(_to).transfer(reward);
+		}
+	}
+
+	function getReward() external nonReentrant {
+			_getReward(msg.sender);
+	}
+
+	function _distributeGivebackAndQueryOracle(uint256 _amount) internal {
+		if (_amount == 0) return;
+		uint256[] memory faucetAddresses = spendingRateOracle.getFaucets();
+		for (uint256 i = 0; i < faucetAddresses.length; i++) {
+			address faucetToken = spendingRateOracle.getFaucetTokenAddress(faucetAddresses[i]);
+			uint256 targetBalance = spendingRateOracle.getFaucetTargetBalance(faucetAddresses[i]);
+			if (faucetToken == address(0)) {
+				uint256 balancesDifference;
+				if (faucetToken.balance < targetBalance) {
+					balancesDifference = targetBalance - faucetToken.balance;
+					if (_amount < balancesDifference) break;
+					_amount -= balancesDifference;
+					payable(faucetAddresses[i]).transfer(balancesDifference);
+				}
+			} else {
+				uint256 balancesDifference;
+				uint256 actualBalance = IERC20(faucetToken).balanceOf(faucetAddresses[i]);
+				if (actualBalance < targetBalance) {
+					balancesDifference = targetBalance - actualBalance;
+					if (_amount < balancesDifference) break;
+					_amount -= balancesDifference;
+					IERC20(faucetToken).safeTransfer(faucetAddresses[i], balancesDifference);
+				}
+			}
+		}
+	}
+
+	function _distributeToUBIAndCommunityPoolAndQueryOracle(uint256 _ubiAmount, uint256 _communityPoolAmount) internal {
+		if (_ubiAmount == 0 || _communityPoolAmount == 0) return;
+		communityPoolBalance += _communityPoolAmount;
+		spendingRateOracle.queryBalance(
+			address(this),
+			communityPoolBalance,
+			address(0);
+		);
```

Comment:
what is this?

Reply:
Querying the community pool balance in the oracle. But yeah, don't think now that is useful. It won't be in prototype.